web/src/settings_users.ts

```diff
@@ -0,0 +1,729 @@
+}
+
+function sort_bot_owner(a: BotInfo, b: BotInfo): number {
+    function owner_name(bot: BotInfo): string {
+        return (bot.bot_owner_full_name ?? "").toLowerCase();
+    }
+
+    return user_sort.compare_a_b(owner_name(a), owner_name(b));
+}
+
+function sort_last_active(a: people.User, b: people.User): number {
+    return user_sort.compare_a_b(
+        presence.last_active_date(a.user_id)?.getTime() ?? 0,
+        presence.last_active_date(b.user_id)?.getTime() ?? 0,
+    );
+}
+
+function get_user_info_row(user_id: number): JQuery {
+    return $(`tr.user_row[data-user-id='${CSS.escape(user_id.toString())}']`);
+}
+
+export function allow_sorting_deactivated_users_list_by_email(): boolean {
+    const deactivated_users = people.get_non_active_realm_users();
+    const deactivated_humans_with_visible_email = deactivated_users.filter(
+        (user) => !user.is_bot && user.delivery_email,
+    );
+
+    return deactivated_humans_with_visible_email.length !== 0;
+}
+
+export function update_view_on_deactivate(user_id: number): void {
+    const $row = get_user_info_row(user_id);
+    if ($row.length === 0) {
+        return;
+    }
+
+    const $button = $row.find("button.deactivate");
+    $button.prop("disabled", false);
+    $row.find("i.deactivated-user-icon").show();
+    $button.addClass("btn-warning reactivate");
+    $button.removeClass("deactivate btn-danger");
+    $button.empty().append($("<i>").addClass(["fa", "fa-user-plus"]).attr("aria-hidden", "true"));
+    $row.removeClass("active-user");
+    $row.addClass("deactivated_user");
+
+    should_redraw_active_users_list = true;
+    should_redraw_deactivated_users_list = true;
+}
+
+export function update_view_on_reactivate(user_id: number): void {
+    const $row = get_user_info_row(user_id);
+    if ($row.length === 0) {
+        return;
+    }
+
+    const $button = $row.find("button.reactivate");
+    $row.find("i.deactivated-user-icon").hide();
+    $button.addClass("btn-danger deactivate");
+    $button.removeClass("btn-warning reactivate");
+    $button.empty().append($("<i>").addClass(["fa", "fa-user-times"]).attr("aria-hidden", "true"));
+    $row.removeClass("deactivated_user");
+    $row.addClass("active-user");
+
+    should_redraw_active_users_list = true;
+    should_redraw_deactivated_users_list = true;
+}
+
+function failed_listing_users(): void {
+    loading.destroy_indicator($("#subs_page_loading_indicator"));
+    const user_id = people.my_current_user_id();
+    blueslip.error("Error while listing users for user_id", {user_id});
+}
+
+function add_value_to_filters(
+    section: ActiveSection | DeactivatedSection,
+    key: keyof SearchFilters,
+    value: number | string,
+): void {
+    if (key === "role_code") {
+        assert(typeof value === "number");
+        section.filters[key] = value;
+    } else {
+        assert(typeof value === "string");
+        section.filters[key] = value;
+    }
```

Comment:
This kind of pattern really wants to use generics.

```suggestion
function add_value_to_filters<Key extends keyof SearchFilters>(
    section: ActiveSection | DeactivatedSection,
    key: Key,
    value: SearchFilters[Key],
): void {
    section.filters[key] = value;
```

Reply:
Done!!